File: Python/Binary Trees/CheckBalancedTree.py

class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def treeHeight(root):
    if root is None:
        return 0
    height = 0
    q = [(root, 0)]

    while q:
       node, level = q.pop(0)
       height = max(height, level)
       if node.left:
           q.append((node.left, level+1))
       if node.right:
           q.append((node.right, level+1))
    return height + 1

def isBalancedTree(root):
    if root is None:
        return True
    leftTree = isBalancedTree(root.left)
    rightTree = isBalancedTree(root.right)

    value = abs(treeHeight(root.left) - treeHeight(root.right)) <= 1

    if(leftTree and rightTree and value):
        return True
    else:
        return False
    
def isBalancedTreeOptimized(root):
    if root is None:
        return (True, 0)
    leftTree = isBalancedTreeOptimized(root.left)
    rightTree = isBalancedTreeOptimized(root.right)

    leftBalanced = leftTree[0]
    rightBalanced = rightTree[0]
    value = abs(leftTree[1] - rightTree[1]) <= 1
   
    isBalanced = leftBalanced and rightBalanced and value
    height = max(leftTree[1], rightTree[1]) + 1

    return (isBalanced, height)


if __name__ == "__main__":
    root = Node(1)
    root.left = Node(2)
    root.right = Node(3)
    root.left.left = Node(4)
    root.left.left.left = Node(5)

    balance, height = isBalancedTreeOptimized(root)

    if balance:
        print('Tree is balanced')
    else:
        print('Tree is not balanced')

File: C++/BInary Trees/CheckBalancedTree.cpp

#include <iostream>
#include<bits/stdc++.h>
using namespace std;

class Node
{
public:
    int data;
    Node *left;
    Node *right;

    Node(int val)
    {
        data = val;
        left = nullptr;
        right = nullptr;
    }
};

void deleteTree(Node *root)
{
    if (root == nullptr)
        return;

    deleteTree(root->left);
    deleteTree(root->right);
    delete root; // Delete the current node AFTER deleting children
}

int Height(Node *root)
{
    if (root == nullptr)
        return 0;
    int left = Height(root->left);
    int right = Height(root->right);
    int ans = max(left, right) + 1;
    return ans;
}

bool isBalancedTree(Node* root){
    if (root == nullptr) return true;

    bool left = isBalancedTree(root->left);
    bool right = isBalancedTree(root->right);

    bool value = abs(Height(root->left) - Height(root->right)) <= 1;

    if(left && right && value) return true;
    else return false;   
}

pair<bool,int> isBalancedOptimized(Node* root){
    if (root == nullptr) return {true, 0};
    pair<bool, int> left = isBalancedOptimized(root->left);
    pair<bool, int> right = isBalancedOptimized(root->right);

    bool leftAns = left.first;
    bool rightAns = right.first;

    bool value = abs(left.second - right.second) <= 1;
    int height = max(left.second, right.second) + 1;

    if(leftAns && rightAns && value) return {true, height};
    else return{false, height};
}

int main(){

    Node *root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
    // root->left->left = new Node(4);
    // root->left->right = new Node(5);
    // root->right->right = new Node(6);
    // root->left->right->left = new Node(7);
    // root->right->right->left = new Node(8);
    // root->right->right->left->left = new Node(9);

    // bool balanced = isBalancedTree(root);
    bool isBalanced = isBalancedOptimized(root).first;

    if(isBalanced) cout<<"Tree is Balanced";
    else cout << "Tree is not balanced";

    deleteTree(root);

    return 0;
}